
ЛАБОРАТОРНАЯ РАБОТА 9. ИСПОЛЬЗОВАНИЕ
РАЗРАБОТАННОГО ПАЙПЛАЙНА ДЛЯ МНОГОМЕРНОЙ
РЕГРЕССИИ
Цели и задачи
Цель лабораторной работы:
научиться применять разработанный пайплайн для тиражирования кода с целью решения широкого круга задач машинного обучения.
Основные задачи:
- получение навыков рефакторинга кода в проектах машинного
обучения;
- получение навыков определения ключевых признаков в задачах машинного обучения;
- получение навыков реализации ключевых стратегий оптимизации моделей регрессии.

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from google.colab import files


In [ ]:
dataset = pd.read_csv('50_Startups.csv')
dataset.head()

R&D Spend  Administration  Marketing Spend       State     Profit
0  165349.20       136897.80        471784.10    New York  192261.83
1  162597.70       151377.59        443898.53  California  191792.06
2  153441.51       101145.55        407934.54     Florida  191050.39
3  144372.41       118671.85        383199.62    New York  182901.99
4  142107.34        91391.77        366168.42     Florida  166187.94

In [ ]:
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, 4].values
print("Матрица признаков"); print(X[:5])
print("Зависимая переменная"); print(y[:5])

Матрица признаков
[[165349.2 136897.8 471784.1 'New York']
 [162597.7 151377.59 443898.53 'California']
 [153441.51 101145.55 407934.54 'Florida']
 [144372.41 118671.85 383199.62 'New York']
 [142107.34 91391.77 366168.42 'Florida']]
Зависимая переменная
[192261.83 191792.06 191050.39 182901.99 166187.94]


In [ ]:
from sklearn.impute import SimpleImputer
# Создание объекта SimpleImputer
imputer = SimpleImputer(missing_values=np.nan, strategy='mean')
# Применение SimpleImputer для заполнения пропусков
X[:, 1:3] = imputer.fit_transform(X[:, 1:3])
# Печать обработанного массива X
print(X)


[[165349.2 136897.8 471784.1 'New York']
 [162597.7 151377.59 443898.53 'California']
 [153441.51 101145.55 407934.54 'Florida']
 [144372.41 118671.85 383199.62 'New York']
 [142107.34 91391.77 366168.42 'Florida']
 [131876.9 99814.71 362861.36 'New York']
 [134615.46 147198.87 127716.82 'California']
 [130298.13 145530.06 323876.68 'Florida']
 [120542.52 148718.95 311613.29 'New York']
 [123334.88 108679.17 304981.62 'California']
 [101913.08 110594.11 229160.95 'Florida']
 [100671.96 91790.61 249744.55 'California']
 [93863.75 127320.38 249839.44 'Florida']
 [91992.39 135495.07 252664.93 'California']
 [119943.24 156547.42 256512.92 'Florida']
 [114523.61 122616.84 261776.23 'New York']
 [78013.11 121597.55 264346.06 'California']
 [94657.16 145077.58 282574.31 'New York']
 [91749.16 114175.79 294919.57 'Florida']
 [86419.7 153514.11 0.0 'New York']
 [76253.86 113867.3 298664.47 'California']
 [78389.47 153773.43 299737.29 'New York']
 [73994.56 122782.75 303319.26 'Florida']
 [67532

In [ ]:
from sklearn.preprocessing import LabelEncoder
Labelencoder_y = LabelEncoder()
print("Зависимая переменная до обработки")
print(y)
print(y)
y = Labelencoder_y.fit_transform(y)
print("Зависимая переменная после обработки")

Зависимая переменная до обработки
[192261.83 191792.06 191050.39 182901.99 166187.94 156991.12 156122.51
 155752.6  152211.77 149759.96 146121.95 144259.4  141585.52 134307.35
 132602.65 129917.04 126992.93 125370.37 124266.9  122776.86 118474.03
 111313.02 110352.25 108733.99 108552.04 107404.34 105733.54 105008.31
 103282.38 101004.64  99937.59  97483.56  97427.84  96778.92  96712.8
  96479.51  90708.19  89949.14  81229.06  81005.76  78239.91  77798.83
  71498.49  69758.98  65200.33  64926.08  49490.75  42559.73  35673.41
  14681.4 ]
[192261.83 191792.06 191050.39 182901.99 166187.94 156991.12 156122.51
 155752.6  152211.77 149759.96 146121.95 144259.4  141585.52 134307.35
 132602.65 129917.04 126992.93 125370.37 124266.9  122776.86 118474.03
 111313.02 110352.25 108733.99 108552.04 107404.34 105733.54 105008.31
 103282.38 101004.64  99937.59  97483.56  97427.84  96778.92  96712.8
  96479.51  90708.19  89949.14  81229.06  81005.76  78239.91  77798.83
  71498.49  69758.98  65200.33  6

In [ ]:
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
from sklearn.compose import ColumnTransformer
# Преобразование столбца с помощью LabelEncoder
labelencoder = LabelEncoder()
X[:, 3] = labelencoder.fit_transform(X[:, 3])
# Преобразование категориального признака с помощью OneHotEncoder
column_transformer = ColumnTransformer(
    transformers=[
        ('encoder', OneHotEncoder(), [3])  # Указываем индекс столбца для кодирования
    ],
    remainder='passthrough'  # Остальные столбцы оставляем без изменений
)
# Применение преобразования
X = column_transformer.fit_transform(X)
# Преобразование в массив
X = np.array(X)
# Вывод результата
print("Перекодировка категориального признака")
print(X[:4, :])

Перекодировка категориального признака
[[0.0 0.0 1.0 165349.2 136897.8 471784.1]
 [1.0 0.0 0.0 162597.7 151377.59 443898.53]
 [0.0 1.0 0.0 153441.51 101145.55 407934.54]
 [0.0 0.0 1.0 144372.41 118671.85 383199.62]]


In [ ]:
X = X[:, 1:]
print(X[:4,:])

[[0.0 1.0 165349.2 136897.8 471784.1]
 [0.0 0.0 162597.7 151377.59 443898.53]
 [1.0 0.0 153441.51 101145.55 407934.54]
 [0.0 1.0 144372.41 118671.85 383199.62]]


In [ ]:
#from sklearn.cross_validation import train_test_split
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

In [ ]:
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train, y_train)

LinearRegression()

In [ ]:
y_pred = regressor.predict(X_test)
print(y_pred)

[22.73971792 31.79851398 32.76702713 10.17872713 48.95244293 25.07783798
  7.62074212 20.43498693 25.32328448 45.16861644]


In [ ]:
import statsmodels.api as sm
# Преобразование данных в тип float
X = X.astype(float)
# Добавляем единичный столбец для константы
X = np.append(arr=np.ones((50, 1)).astype(int), values=X, axis=1)
# Указание доступных индексов
X_opt = X[:, [0, 1, 2, 3, 4, 5]]  # Обновите индексы, чтобы они находились в пределах диапазона
regressor_OLS = sm.OLS(endog=y, exog=X_opt).fit()
regressor_OLS.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.960
Model:                            OLS   Adj. R-squared:                  0.956
Method:                 Least Squares   F-statistic:                     212.6
Date:                Tue, 17 Dec 2024   Prob (F-statistic):           1.22e-29
Time:                        19:49:50   Log-Likelihood:                -123.78
No. Observations:                  50   AIC:                             259.6
Df Residuals:                      44   BIC:                             271.0
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          1.2027      2.237      0.538      0.594      -3.306       5.711
x1             0.1929      1.095      0.176      0.861      -2.015       2.400
x2            -0.2364      1.058     -0.223      0.824      -2.369       1.896
x3             0.0003   1.51e-05     19.988      0.000       0.000       0.000
x4          5.233e-07    1.7e-05      0.031      0.976   -3.37e-05    3.47e-05
x5          4.897e-06   5.57e-06      0.879      0.384   -6.33e-06    1.61e-05
==============================================================================
Omnibus:                        3.294   Durbin-Watson:                   1.285
Prob(Omnibus):                  0.193   Jarque-Bera (JB):                3.155
Skew:                           0.582   Prob(JB):                        0.206
Kurtosis:                       2.602   Cond. No.                     1.45e+06
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.45e+06. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [ ]:
X_opt = X[:, [0, 1, 3, 4, 5]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.960
Model:                            OLS   Adj. R-squared:                  0.957
Method:                 Least Squares   F-statistic:                     271.5
Date:                Tue, 17 Dec 2024   Prob (F-statistic):           7.07e-31
Time:                        19:49:59   Log-Likelihood:                -123.81
No. Observations:                  50   AIC:                             257.6
Df Residuals:                      45   BIC:                             267.2
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          1.0949      2.161      0.507      0.615      -3.258       5.448
x1             0.3135      0.943      0.332      0.741      -1.586       2.213
x2             0.0003   1.49e-05     20.236      0.000       0.000       0.000
x3          5.737e-07   1.68e-05      0.034      0.973   -3.32e-05    3.44e-05
x4          4.897e-06   5.51e-06      0.889      0.379    -6.2e-06     1.6e-05
==============================================================================
Omnibus:                        3.623   Durbin-Watson:                   1.273
Prob(Omnibus):                  0.163   Jarque-Bera (JB):                3.460
Skew:                           0.623   Prob(JB):                        0.177
Kurtosis:                       2.669   Cond. No.                     1.40e+06
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.4e+06. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [ ]:
X_opt = X[:, [0, 3, 4, 5]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.960
Model:                            OLS   Adj. R-squared:                  0.958
Method:                 Least Squares   F-statistic:                     369.1
Date:                Tue, 17 Dec 2024   Prob (F-statistic):           3.56e-32
Time:                        19:50:19   Log-Likelihood:                -123.87
No. Observations:                  50   AIC:                             255.7
Df Residuals:                      46   BIC:                             263.4
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          1.1176      2.139      0.522      0.604      -3.188       5.423
x1             0.0003   1.47e-05     20.469      0.000       0.000       0.000
x2          8.292e-07   1.66e-05      0.050      0.960   -3.26e-05    3.43e-05
x3           5.25e-06   5.35e-06      0.981      0.332   -5.53e-06     1.6e-05
==============================================================================
Omnibus:                        3.506   Durbin-Watson:                   1.263
Prob(Omnibus):                  0.173   Jarque-Bera (JB):                3.330
Skew:                           0.613   Prob(JB):                        0.189
Kurtosis:                       2.692   Cond. No.                     1.40e+06
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.4e+06. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [ ]:
X_opt = X[:, [0, 3]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.959
Model:                            OLS   Adj. R-squared:                  0.958
Method:                 Least Squares   F-statistic:                     1129.
Date:                Tue, 17 Dec 2024   Prob (F-statistic):           5.25e-35
Time:                        19:50:44   Log-Likelihood:                -124.43
No. Observations:                  50   AIC:                             252.9
Df Residuals:                      48   BIC:                             256.7
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          1.5703      0.802      1.959      0.056      -0.042       3.182
x1             0.0003   9.26e-06     33.599      0.000       0.000       0.000
==============================================================================
Omnibus:                        3.750   Durbin-Watson:                   1.294
Prob(Omnibus):                  0.153   Jarque-Bera (JB):                3.598
Skew:                           0.633   Prob(JB):                        0.165
Kurtosis:                       2.648   Cond. No.                     1.65e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.65e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Индивидуальное задание

In [ ]:
data = pd.read_csv('bodyfat.csv')
data.head()

Density  BodyFat  Age  Weight  Height  Neck  Chest  Abdomen    Hip  Thigh  \
0   1.0708     12.3   23  154.25   67.75  36.2   93.1     85.2   94.5   59.0   
1   1.0853      6.1   22  173.25   72.25  38.5   93.6     83.0   98.7   58.7   
2   1.0414     25.3   22  154.00   66.25  34.0   95.8     87.9   99.2   59.6   
3   1.0751     10.4   26  184.75   72.25  37.4  101.8     86.4  101.2   60.1   
4   1.0340     28.7   24  184.25   71.25  34.4   97.3    100.0  101.9   63.2   

   Knee  Ankle  Biceps  Forearm  Wrist  
0  37.3   21.9    32.0     27.4   17.1  
1  37.3   23.4    30.5     28.9   18.2  
2  38.9   24.0    28.8     25.2   16.6  
3  37.3   22.8    32.4     29.4   18.2  
4  42.2   24.0    32.2     27.7   17.7

In [ ]:
import statsmodels.api as sm
import pandas as pd

# Загрузка и подготовка данных
#  X - признаки, y - целевая переменная (BodyFat)
X = data.drop('BodyFat', axis=1)  # Признаки
y = data['BodyFat']  # Целевая переменная

In [ ]:
# Добавляем столбец единиц для свободного члена
X = sm.add_constant(X)

# Строим модель линейной регрессии
model = sm.OLS(y, X).fit()

# Выводим результаты модели
print(model.summary())

# Применяем Backward Elimination
# Задаем порог для p-уровня
p_threshold = 0.05

                            OLS Regression Results                            
Dep. Variable:                BodyFat   R-squared:                       0.978
Model:                            OLS   Adj. R-squared:                  0.977
Method:                 Least Squares   F-statistic:                     756.3
Date:                Wed, 18 Dec 2024   Prob (F-statistic):          8.50e-188
Time:                        22:11:51   Log-Likelihood:                -410.93
No. Observations:                 252   AIC:                             851.9
Df Residuals:                     237   BIC:                             904.8
Df Model:                          14                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        450.0126     10.713     42.005      0.0

In [ ]:
# Процесс исключения признаков
while True:
    # Находим наибольший p-уровень среди признаков
    p_values = model.pvalues
    max_p_value = p_values.max()

    # Если наибольший p-уровень больше порога, удаляем признак с этим p-уровнем
    if max_p_value > p_threshold:
        feature_with_max_p = p_values.idxmax()  # Признак с максимальным p-уровнем
        X = X.drop(columns=[feature_with_max_p])  # Удаляем его из данных
        model = sm.OLS(y, X).fit()  # Строим модель с оставшимися признаками
    else:
        break

In [ ]:
# Результаты финальной модели
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                BodyFat   R-squared:                       0.977
Model:                            OLS   Adj. R-squared:                  0.977
Method:                 Least Squares   F-statistic:                     5341.
Date:                Wed, 18 Dec 2024   Prob (F-statistic):          3.27e-205
Time:                        22:12:47   Log-Likelihood:                -415.93
No. Observations:                 252   AIC:                             837.9
Df Residuals:                     249   BIC:                             848.4
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        455.5088      7.035     64.747      0.0

Итог:
Модель объясняет более 97% вариации целевой переменной, что является очень хорошим результатом.
Признаки "плотность" и "окружность грудной клетки" имеют значительное влияние на содержание жира в организме.

 Вывод:
 модель OLS (Ordinary Least Squares),  метод наименьших квадратов для линейной регрессии.
Основные показатели:
Dep. Variable (Целевая переменная): BodyFat — переменная, которую мы пытаемся предсказать (процент содержания жира в организме).
R-squared (R²): 0.977
Это коэффициент детерминации, который показывает, какую долю вариативности целевой переменной модель объясняет. Значение 0.977 означает, что модель объясняет 97.7% вариации в содержании жира в организме на основе выбранных признаков. Это указывает на очень хорошую модель.
Adj. R-squared (Отрегулированный R²): 0.977
Этот показатель также близок к R², но учитывает количество признаков в модели. Он используется для сравнения моделей с разным количеством признаков.
F-statistic (F-статистика): 5341.0
Это статистика, которая используется для проверки значимости всей модели. Высокое значение F-статистики указывает на то, что модель значимо предсказывает целевую переменную.
Prob (F-statistic): 3.27e-205
Это p-уровень для F-статистики. Если p-уровень меньше 0.05 (или в нашем случае меньше 0.001), это говорит о том, что модель в целом статистически значима.
Log-Likelihood (Логарифм правдоподобия): -415.93
Используется для оценки качества модели. Чем выше это значение, тем лучше модель подходит к данным.
AIC (Akaike Information Criterion): 837.9 и BIC (Bayesian Information Criterion): 848.4
Эти критерии используются для оценки модели. Чем ниже значения AIC и BIC, тем лучше модель. Это важно для сравнения разных моделей.
Признаки и их коэффициенты:
const (Свободный член): 455.5088
Это значение предполагаемой величины для целевой переменной BodyFat, когда все признаки равны нулю. В данном случае это значение очень большое, что может указывать на проблемы с масштабированием данных или неправильное использование признаков.
Density (Плотность): -418.4199
Это коэффициент для признака "плотность" в модели. Это означает, что на каждый 1 ед. изменения плотности содержание жира в организме будет изменяться на -418.42 единиц. Отрицательное значение указывает на отрицательную зависимость: с увеличением плотности содержание жира в организме снижается.
Chest (Окружность грудной клетки): 0.0527
Это коэффициент для признака "окружность грудной клетки". Значение говорит о том, что с увеличением окружности грудной клетки содержание жира в организме увеличивается на 0.0527 единиц.
Статистические показатели для каждого признака:
coef (Коэффициент): Значения коэффициентов для признаков. Они показывают, насколько сильно каждый признак влияет на целевую переменную.
std err (Стандартная ошибка): Измеряет точность оценки коэффициента. Чем меньше стандартная ошибка, тем точнее оценка.
t: Статистика t для проверки значимости коэффициента (t-статистика). Чем выше значение t, тем значимее признак для модели.
P>|t|: p-уровень для каждого признака. Если p-уровень меньше 0.05 (или 0.01), то признак считается статистически значимым.
Для всех признаков в модели p-уровни равны нулю, что подтверждает их значимость.
[0.025, 0.975]: 95% доверительный интервал для каждого коэффициента. Это диапазон, в котором с вероятностью 95% находится истинное значение коэффициента.

Omnibus и Prob(Omnibus): Статистика теста на нормальность остатков модели. Низкое значение p-уровня (0.000) говорит о том, что остатки не распределены нормально, что может свидетельствовать о проблемах модели.

Durbin-Watson: 2.063

Этот показатель проверяет наличие автокорреляции (корреляции между остатками модели). Значение близкое к 2 говорит о том, что автокорреляции нет.
Jarque-Bera (JB) и Prob(JB): Тест на нормальность остатков. Высокое значение JB и p-уровень 0.00 говорят о том, что остатки не нормально распределены, что может требовать дополнительных проверок или коррекций.

Skew и Kurtosis: Показатели асимметрии и куртоза (острота распределения остатков). В данном случае значения слишком высоки, что также подтверждает, что остатки не имеют нормальное распределение.

Condition number: 1.15e+04

Этот показатель указывает на возможные проблемы с мультиколлинеарностью (взаимосвязью между признаками). Значение выше 10000 может означать наличие сильной мультиколлинеарности, что может ухудшать качество модели.